src/zarr/core/metadata/v3.py

```diff
@@ -445,6 +448,7 @@ class DataType(Enum):
     float64 = "float64"
     complex64 = "complex64"
     complex128 = "complex128"
+    string = "string"
```

Comment:
Is this the value that shows up in `zarr.json`? If so I think we should be a bit careful here to not claim this as *the* string data type. Maybe call this `vlen-utf8`?

As I write this, I realize I don't really understand the relationship between what exactly effect `data_type` has, so maybe this isn't so important.

Reply:
> From my point of view, the dtype says what sort of value the data should be cast to after all the codecs are done decoding.

Thanks, this is the bit I was missing earlier.

One perspective I like is Apache Parquet's distinction between [Logical Types](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md) and physical ("primitive") Types. I think in this case our `data_type` field is logical, and so `string` is appropriate. Does that sound right?

Reply:
I think Parquet's framework is useful. But it feels a bit unclear how it applies in Zarr. From the point of view of a codec pipeline, the only physical datatype is `bytes`, right? We basically have `in_memory_data -> [encoding pipeline] -> bytes_to_store`? At what point is data translated from logical to physical?

Bringing Xarray into the picture makes it even harder, because Xarray does its own coding (e.g. from int to datetime).

I think a "logical type" framework like this would help Zarr a lot, but it feels like a big project to fully implement.

Reply:
@rabernat
> Also aware the we have some ugly coupling between a codec and a type.

Re: https://github.com/zarr-developers/zeps/pull/47#issuecomment-2238480835.

I still think it is worth considering the consolidation of `vlen-utf8`, `vlen-bytes`, `vlen-array` into a new codec name in Zarr V3 metadata to fix this coupling. These codecs all encode the data the same, just interleaved lengths and bytes.

I think `vlen_v2` is a good candidate name to indicate it is legacy and discourage use in Zarr V3[^1], similarly to the `v2` chunk key encoding. In `zarrs`, I use the same codec implementation whether it is a V2 array with `vlen-*` or a V3 array with `vlen_v2`, it is just a [metadata change](https://github.com/LDeakin/zarrs/blob/10d3b80e0fbf98515eb14e4197f011d0301ec196/zarrs_metadata/src/v2_to_v3.rs#L148-L154).

> Lachlan, what would the metadata have to look like to be interoperable with your zarrs implmentation?

It is compatible now with a change in [runtime config](https://github.com/LDeakin/zarrs/blob/10d3b80e0fbf98515eb14e4197f011d0301ec196/zarrs/src/config.rs#L126). I can change the default depending on what gets spec'd.

[^1]: Out of the scope of this PR, but I hope `zarr-python` can get a codec like what @jbms proposed [here](https://github.com/zarr-developers/zeps/pull/47#issuecomment-1710505141) (post v3.0.0?) as the preferred `vlen` codec going forward. Unlike the V2 `vlen-*` codecs, it can support partial decoding. I have experimental support for such a codec in `zarrs`.